File: src/api/aria2.ts
/**
 * @fileoverview Aria2 API — invoke() transport layer.
 *
 * All aria2 RPC calls go through Tauri invoke() to the Rust backend.
 * The Rust Aria2Client handles HTTP JSON-RPC communication with Aria2 Next.
 */
import { invoke } from '@tauri-apps/api/core'
import { changeKeysToCamelCase, formatOptionsForEngine } from '@shared/utils'
import type {
  Aria2Task,
  Aria2RawGlobalStat,
  Aria2EngineOptions,
  Aria2File,
  AppConfig,
  ExternalDownloadContext,
} from '@shared/types'
import { formatLogFields, logger } from '@shared/logger'
import { resolveDownloadDir } from '@shared/utils/fileCategory'
import { sanitizeAria2OutHint } from '@shared/utils/batchHelpers'

/**
 * Engine readiness state.
 * With the Rust backend transport, readiness is determined by the engine
 * lifecycle commands — the Aria2Client is always available once credentials
 * are set by `on_engine_ready`.
 */
let engineReady = false

/** Returns true when the aria2 engine has started and is accepting RPC. */
export function isEngineReady(): boolean {
  return engineReady
}

/** Marks the engine as ready/unready. */
export function setEngineReady(ready: boolean): void {
  engineReady = ready
}

/** Retrieves aria2 engine version and list of enabled features. */
export async function getVersion(): Promise<{ version: string; enabledFeatures: string[] }> {
  return invoke<{ version: string; enabledFeatures: string[] }>('aria2_get_version')
}

/** Fetches all global aria2 configuration options as camelCase keys. */
export async function getGlobalOption(): Promise<Record<string, string>> {
  const data = await invoke<Record<string, string>>('aria2_get_global_option')
  return changeKeysToCamelCase(data) as Record<string, string>
}

/** Fetches aggregated download/upload statistics from aria2. */
export async function getGlobalStat(): Promise<Aria2RawGlobalStat> {
  return invoke<Aria2RawGlobalStat>('aria2_get_global_stat')
}

/** Updates aria2 global configuration at runtime. */
export async function changeGlobalOption(options: Partial<AppConfig>): Promise<void> {
  const engineOptions = formatOptionsForEngine(options)
  logger.debug('aria2.changeGlobalOption', engineOptions)
  await invoke<string>('aria2_change_global_option', { options: engineOptions })
}

/** Fetches the option set for a specific download task as camelCase keys. */
export async function getOption(params: { gid: string }): Promise<Record<string, string>> {
  const data = await invoke<Record<string, string>>('aria2_get_option', { gid: params.gid })
  return changeKeysToCamelCase(data) as Record<string, string>
}

/** Modifies options for a specific download task at runtime. */
export async function changeOption(params: { gid: string; options: Aria2EngineOptions }): Promise<void> {
  const engineOptions = formatOptionsForEngine(params.options)
  await invoke<string>('aria2_change_option', { gid: params.gid, options: engineOptions })
}

/** Retrieves the file list for a download task by GID. */
export async function getFiles(params: { gid: string }): Promise<Aria2File[]> {
  const data = await invoke<Record<string, unknown>[]>('aria2_get_files', { gid: params.gid })
  return data.map((f) => changeKeysToCamelCase(f)) as unknown as Aria2File[]
}

/** Fetches only active tasks (no waiting). */
export async function fetchActiveTaskList(): Promise<Aria2Task[]> {
  return invoke<Aria2Task[]>('aria2_fetch_active_task_list')
}

/** Fetches task list by status type: active+waiting or stopped. */
export async function fetchTaskList(params: { type: string; limit?: number }): Promise<Aria2Task[]> {
  return invoke<Aria2Task[]>('aria2_fetch_task_list', {
    type: params.type,
    limit: params.limit ?? null,
  })
}

/** Fetches a single task's full status by GID. */
export async function fetchTaskItem(params: { gid: string }): Promise<Aria2Task> {
  return invoke<Aria2Task>('aria2_fetch_task_item', { gid: params.gid })
}

/** Adds one or more URI downloads with per-URI output filename overrides. */
export async function addUri(params: {
  uris: string[]
  outs: string[]
  options: Aria2EngineOptions
  fileCategory?: {
    enabled: boolean
    categories: import('@shared/types').FileCategory[]
    contexts?: Record<string, ExternalDownloadContext>
  }
}): Promise<string[]> {
  const { uris, outs, options, fileCategory } = params
  const engineOptions = formatOptionsForEngine(options)

  // Each URI gets its own aria2 task with optional per-URI overrides
  const tasks = uris.map(async (uri, index) => {
    const opts: Record<string, string> = { ...engineOptions }
    if (outs[index]) opts.out = outs[index]

    // Defense-in-depth: sanitize out for filesystem safety (#261, #264).
    // Rust sanitize_out_option is the authoritative boundary; this is belt-and-suspenders.
    if (opts.out) opts.out = sanitizeAria2OutHint(opts.out)
    if (!opts.out) delete opts.out

    // Smart file classification: resolve per-URI download directory
    if (fileCategory?.enabled && fileCategory.categories.length > 0) {
      const context = fileCategory.contexts?.[uri]
      opts.dir = resolveDownloadDir(opts.out || uri, opts.dir || '', true, fileCategory.categories, {
        urls: [uri, context?.finalUrl ?? '', context?.url ?? '', context?.referer ?? ''],
      })
    }

    return invoke<string>('aria2_add_uri', { uris: [uri], options: opts })
  })

  const gids = await Promise.all(tasks)
  logger.info(
    'aria2.addUri',
    formatLogFields({
      added: gids.length,
      gids: `[${gids.join(',')}]`,
    }),
  )
  return gids
}

/**
 * Adds a single download with all URIs as mirrors (alternative sources).
 */
export async function addUriAtomic(params: { uris: string[]; options: Aria2EngineOptions }): Promise<string> {
  const { uris, options } = params
  const engineOptions = formatOptionsForEngine(options)
  const gid = await invoke<string>('aria2_add_uri', { uris, options: engineOptions })
  logger.debug('aria2.addUriAtomic', `gid=${gid} mirrors=${uris.length}`)
  return gid
}

/** Forcefully removes a download task by GID. */
export async function removeTask(params: { gid: string }): Promise<string> {
  return invoke<string>('aria2_force_remove', { gid: params.gid })
}

/** Forcefully pauses a download task by GID. */
export async function forcePauseTask(params: { gid: string }): Promise<string> {
  return invoke<string>('aria2_force_pause', { gid: params.gid })
}

/** Pauses a download task by GID (graceful). */
export async function pauseTask(params: { gid: string }): Promise<string> {
  return invoke<string>('aria2_pause', { gid: params.gid })
}

/** Resumes a paused download task by GID. */
export async function resumeTask(params: { gid: string }): Promise<string> {
  return invoke<string>('aria2_unpause', { gid: params.gid })
}

/** Saves the current aria2 session to disk. */
export async function saveSession(): Promise<string> {
  return invoke<string>('aria2_save_session')
}

/** Removes a completed/errored task record from the download list. */
export async function removeTaskRecord(params: { gid: string }): Promise<string> {
  return invoke<string>('aria2_remove_download_result', { gid: params.gid })
}

/** Purges all completed/errored task records from the download list. */
export async function purgeTaskRecord(): Promise<string> {
  return invoke<string>('aria2_purge_download_result')
}

/** Batch-resumes multiple tasks by GID array via multicall. */
export async function batchResumeTask(params: { gids: string[] }): Promise<unknown[][]> {
  return invoke<unknown[][]>('aria2_batch_unpause', { gids: params.gids })
}

/** Batch-pauses multiple tasks by GID array via multicall (force). */
export async function batchPauseTask(params: { gids: string[] }): Promise<unknown[][]> {
  return invoke<unknown[][]>('aria2_batch_force_pause', { gids: params.gids })
}

/** Alias for batchPauseTask — force-pauses multiple tasks. */
export async function batchForcePauseTask(params: { gids: string[] }): Promise<unknown[][]> {
  return batchPauseTask(params)
}

/** Batch-removes multiple tasks by GID array via multicall (force). */
export async function batchRemoveTask(params: { gids: string[] }): Promise<unknown[][]> {
  return invoke<unknown[][]>('aria2_batch_force_remove', { gids: params.gids })
}

const api = {
  getVersion,
  getGlobalOption,
  getGlobalStat,
  changeGlobalOption,
  getOption,
  changeOption,
  getFiles,
  fetchActiveTaskList,
  fetchTaskList,
  fetchTaskItem,
  addUri,
  addUriAtomic,
  removeTask,
  forcePauseTask,
  pauseTask,
  resumeTask,
  saveSession,
  removeTaskRecord,
  purgeTaskRecord,
  batchResumeTask,
  batchPauseTask,
  batchForcePauseTask,
  batchRemoveTask,
}

export default api

File: src-tauri/src/commands/aria2.rs
//! Tauri commands exposing aria2 RPC operations to the frontend.
//!
//! These commands serve as the invoke() transport layer. Each command maps
//! to one or more aria2 RPC methods.

use crate::aria2::client::Aria2State;
use crate::aria2::types::{Aria2File, Aria2Task};
use crate::commands::net::decode_filename_encoding;
use crate::error::AppError;
use tauri::State;

/// Fetch task list by type.
#[tauri::command]
pub async fn aria2_fetch_task_list(
    state: State<'_, Aria2State>,
    r#type: String,
    limit: Option<i64>,
) -> Result<Vec<Aria2Task>, AppError> {
    match r#type.as_str() {
        "active" => {
            let (active, waiting) =
                tokio::try_join!(state.0.tell_active(), state.0.tell_waiting(0, 1000),)?;
            let mut result = active;
            result.extend(waiting);
            Ok(result)
        }
        "waiting" => state.0.tell_waiting(0, limit.unwrap_or(1000)).await,
        _ => state.0.tell_stopped(0, limit.unwrap_or(1000)).await,
    }
}

/// Fetch only active tasks (no waiting).
#[tauri::command]
pub async fn aria2_fetch_active_task_list(
    state: State<'_, Aria2State>,
) -> Result<Vec<Aria2Task>, AppError> {
    state.0.tell_active().await
}

/// Fetch a single task's full status by GID.
#[tauri::command]
pub async fn aria2_fetch_task_item(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<Aria2Task, AppError> {
    state.0.tell_status(&gid).await
}

/// Get aria2 engine version and enabled features.
#[tauri::command]
pub async fn aria2_get_version(
    state: State<'_, Aria2State>,
) -> Result<serde_json::Value, AppError> {
    state.0.get_version().await
}

/// Get global aria2 options.
#[tauri::command]
pub async fn aria2_get_global_option(
    state: State<'_, Aria2State>,
) -> Result<serde_json::Value, AppError> {
    state.0.get_global_option().await
}

/// Get global download/upload statistics.
#[tauri::command]
pub async fn aria2_get_global_stat(
    state: State<'_, Aria2State>,
) -> Result<serde_json::Value, AppError> {
    let stat = state.0.get_global_stat().await?;
    serde_json::to_value(&stat).map_err(|e| AppError::Aria2(format!("serialize stat: {e}")))
}

/// Change global aria2 options at runtime.
#[tauri::command]
pub async fn aria2_change_global_option(
    state: State<'_, Aria2State>,
    options: serde_json::Map<String, serde_json::Value>,
) -> Result<String, AppError> {
    state.0.change_global_option(options).await
}

/// Get per-task options.
#[tauri::command]
pub async fn aria2_get_option(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<serde_json::Value, AppError> {
    state.0.get_option(&gid).await
}

/// Change per-task options.
#[tauri::command]
pub async fn aria2_change_option(
    state: State<'_, Aria2State>,
    gid: String,
    options: serde_json::Value,
) -> Result<String, AppError> {
    state.0.change_option(&gid, options).await
}

/// Get file list for a task.
#[tauri::command]
pub async fn aria2_get_files(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<Vec<Aria2File>, AppError> {
    state.0.get_files(&gid).await
}

// ── `out` option sanitization ────────────────────────────────────────

/// Sanitizes an `out` option value into a safe, platform-valid filename.
///
/// aria2's `out` option must be a plain filename relative to `dir`.  aria2
/// itself performs **no** filename sanitization — it passes the value
/// directly to the OS `open()` call.  This function is the authoritative
/// safety boundary.
///
/// Three-step pipeline:
///   1. **Basename extraction** — strips path separators (including Windows
///      drive letters, UNC prefixes, and Unix absolute paths).
///   2. **NUL rejection** — NUL bytes truncate C strings inside aria2.
///   3. **Industry-standard sanitization** via the `sanitize-filename` crate
///      (same character set as Chrome `filename_util.cc` and Node.js
///      `sanitize-filename`):
///      - Replaces `/ \ : * ? " < > |` with `_`
///      - Removes ASCII control chars (0x00–0x1F, 0x7F) and C1 (0x80–0x9F)
///      - Rejects Windows reserved names (CON, NUL, COM1, LPT1, etc.)
///      - Strips trailing dots and spaces (Windows rejects these)
///      - Truncates to 255 bytes (filesystem limit)
///
/// Returns `None` for values that reduce to empty after sanitization.
fn sanitize_out_option(raw: &str) -> Option<String> {
    if raw.is_empty() {
        return None;
    }
    // 1. Basename extraction — split on both separators for cross-platform.
    let basename = raw.rsplit(['/', '\\']).next().unwrap_or(raw);
    if basename.is_empty() || basename == "." || basename == ".." {
        return None;
    }
    // 2. Reject NUL bytes early (truncate C strings inside aria2).
    if basename.contains('\0') {
        return None;
    }
    // 3. Industry-standard sanitization (Chrome / sanitize-filename char set).
    //    Always use Windows rules (most restrictive) regardless of build target
    //    to ensure filenames are safe when the Rust backend runs on any platform
    //    but may serve files destined for Windows clients.
    let decoded = decode_filename_encoding(basename);
    let sanitized = sanitize_filename::sanitize_with_options(
        decoded.as_str(),
        sanitize_filename::Options {
            windows: true,
            truncate: true,
            replacement: "_",
        },
    );
    let result = sanitized.trim().to_string();
    if result.is_empty() {
        return None;
    }
    Some(result)
}

/// Add URI download(s). Each URI gets its own aria2 task with optional
/// per-URI `out` filename override and file-category directory resolution.
#[tauri::command]
pub async fn aria2_add_uri(
    state: State<'_, Aria2State>,
    uris: Vec<String>,
    mut options: serde_json::Value,
) -> Result<String, AppError> {
    // Enforce out = safe-filename invariant before forwarding to aria2.
    // Prevents path traversal (#261) and illegal-character crashes (#264).
    if let Some(opts) = options.as_object_mut() {
        if let Some(out_val) = opts.get("out").and_then(|v| v.as_str()).map(String::from) {
            match sanitize_out_option(&out_val) {
                Some(ref clean) if *clean != out_val => {
                    log::warn!("aria2:add-uri sanitized out: {:?} → {:?}", out_val, clean);
                    opts.insert("out".to_string(), serde_json::Value::String(clean.clone()));
                }
                None => {
                    log::warn!("aria2:add-uri removed invalid out option");
                    opts.remove("out");
                }
                _ => {} // already a clean filename — no action needed
            }
        }
    }
    log::info!("aria2:add-uri count={}", uris.len());
    state.0.add_uri(uris, options).await
}

/// Forcefully remove a task by GID.
#[tauri::command]
pub async fn aria2_force_remove(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<String, AppError> {
    log::info!("aria2:remove gid={gid}");
    state.0.force_remove(&gid).await
}

/// Forcefully pause a task by GID.
#[tauri::command]
pub async fn aria2_force_pause(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<String, AppError> {
    log::debug!("aria2:force-pause gid={gid}");
    state.0.force_pause(&gid).await
}

/// Gracefully pause a task.
#[tauri::command]
pub async fn aria2_pause(state: State<'_, Aria2State>, gid: String) -> Result<String, AppError> {
    log::debug!("aria2:pause gid={gid}");
    state.0.pause(&gid).await
}

/// Resume a paused task.
#[tauri::command]
pub async fn aria2_unpause(state: State<'_, Aria2State>, gid: String) -> Result<String, AppError> {
    log::debug!("aria2:resume gid={gid}");
    state.0.unpause(&gid).await
}

/// Save the current aria2 session to disk.
#[tauri::command]
pub async fn aria2_save_session(state: State<'_, Aria2State>) -> Result<String, AppError> {
    state.0.save_session().await
}

/// Remove a completed/errored task record from aria2's download list.
#[tauri::command]
pub async fn aria2_remove_download_result(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<String, AppError> {
    state.0.remove_download_result(&gid).await
}

/// Purge all completed/errored download results.
#[tauri::command]
pub async fn aria2_purge_download_result(state: State<'_, Aria2State>) -> Result<String, AppError> {
    log::info!("aria2:purge-results");
    state.0.purge_download_result().await
}

/// Batch resume multiple tasks via multicall.
#[tauri::command]
pub async fn aria2_batch_unpause(
    state: State<'_, Aria2State>,
    gids: Vec<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    log::info!("aria2:batch-resume count={}", gids.len());
    let calls = gids
        .into_iter()
        .map(|gid| ("unpause".to_string(), vec![serde_json::Value::String(gid)]))
        .collect();
    state.0.multicall(calls).await
}

/// Batch force-pause multiple tasks via multicall.
#[tauri::command]
pub async fn aria2_batch_force_pause(
    state: State<'_, Aria2State>,
    gids: Vec<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    log::info!("aria2:batch-pause count={}", gids.len());
    let calls = gids
        .into_iter()
        .map(|gid| {
            (
                "forcePause".to_string(),
                vec![serde_json::Value::String(gid)],
            )
        })
        .collect();
    state.0.multicall(calls).await
}

/// Batch force-remove multiple tasks via multicall.
#[tauri::command]
pub async fn aria2_batch_force_remove(
    state: State<'_, Aria2State>,
    gids: Vec<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    log::info!("aria2:batch-remove count={}", gids.len());
    let calls = gids
        .into_iter()
        .map(|gid| {
            (
                "forceRemove".to_string(),
                vec![serde_json::Value::String(gid)],
            )
        })
        .collect();
    state.0.multicall(calls).await
}

#[cfg(test)]
mod tests {
    use super::sanitize_out_option;

    #[test]
    fn bare_filename_passes_through() {
        assert_eq!(sanitize_out_option("file.zip").as_deref(), Some("file.zip"));
    }

    #[test]
    fn windows_backslash_absolute_extracts_basename() {
        assert_eq!(
            sanitize_out_option("C:\\Users\\u\\Downloads\\file.zip").as_deref(),
            Some("file.zip")
        );
    }

    #[test]
    fn forward_slash_absolute_extracts_basename() {
        assert_eq!(
            sanitize_out_option("C:/Users/u/Downloads/file.zip").as_deref(),
            Some("file.zip")
        );
    }

    #[test]
    fn unc_path_extracts_basename() {
        assert_eq!(
            sanitize_out_option("\\\\server\\share\\file.zip").as_deref(),
            Some("file.zip")
        );
    }

    #[test]
    fn parent_traversal_extracts_basename() {
        assert_eq!(
            sanitize_out_option("../evil.exe").as_deref(),
            Some("evil.exe")
        );
    }

    #[test]
    fn dotdot_only_rejected() {
        assert_eq!(sanitize_out_option(".."), None);
    }

    #[test]
    fn dot_only_rejected() {
        assert_eq!(sanitize_out_option("."), None);
    }

    #[test]
    fn empty_rejected() {
        assert_eq!(sanitize_out_option(""), None);
    }

    #[test]
    fn nul_byte_rejected() {
        assert_eq!(sanitize_out_option("file\0.zip"), None);
    }

    #[test]
    fn accented_filename_preserved() {
        assert_eq!(
            sanitize_out_option("C:/Downloads/résumé.zip").as_deref(),
            Some("résumé.zip")
        );
    }

    #[test]
    fn trailing_separator_rejected() {
        assert_eq!(sanitize_out_option("path/to/"), None);
    }

    #[test]
    fn issue_261_regression() {
        assert_eq!(
            sanitize_out_option("C:/Users/37472/Downloads/sysdiag-all-x64.exe").as_deref(),
            Some("sysdiag-all-x64.exe")
        );
    }

    // ── #264: illegal character sanitization ────────────────────────

    #[test]
    fn issue_264_twitter_cdn_filename() {
        // Extension sends "G9v9wWdasAYNqt9?format=jpg&name=large" as filename.
        // `?` is replaced with `_` by the crate; `&` and `=` are legal filename
        // chars and pass through unchanged.
        assert_eq!(
            sanitize_out_option("G9v9wWdasAYNqt9?format=jpg&name=large").as_deref(),
            Some("G9v9wWdasAYNqt9_format=jpg&name=large")
        );
    }

    #[test]
    fn replaces_windows_illegal_chars() {
        assert_eq!(
            sanitize_out_option("a<b>c:d*e.jpg").as_deref(),
            Some("a_b_c_d_e.jpg")
        );
    }

    #[test]
    fn replaces_pipe_and_quotes() {
        assert_eq!(
            sanitize_out_option("file\"|pipe.txt").as_deref(),
            Some("file__pipe.txt")
        );
    }

    #[test]
    fn question_mark_in_filename_replaced() {
        // "what?.jpg" → "what_.jpg" (not truncated to "what")
        assert_eq!(
            sanitize_out_option("what?.jpg").as_deref(),
            Some("what_.jpg")
        );
    }

    #[test]
    fn percent_encoded_rfc2047_out_decodes_before_sanitize() {
        assert_eq!(
            sanitize_out_option("=%3FUTF-8%3FB%3F0JjQotCe0JPQmCDQm9CU0KMgMjAyNi54bHN4%3F=")
                .as_deref(),
            Some("ИТОГИ ЛДУ 2026.xlsx")
        );
    }

    #[test]
    fn percent_encoded_utf8_out_decodes_before_sanitize() {
        assert_eq!(
            sanitize_out_option("K430006866701%20%20%20%20%2020251022%20%20%20ASKO%20%20%20%20CW5937GCN%20%20%20%20%20CW51237GCN%E8%AF%B4%E6%98%8E%E4%B9%A6%28%E6%96%B0%E5%9B%BD%E6%A0%87%29.pdf").as_deref(),
            Some("K430006866701     20251022   ASKO    CW5937GCN     CW51237GCN说明书(新国标).pdf")
        );
    }

    #[test]
    fn percent_encoded_slash_out_stays_single_safe_filename() {
        assert_eq!(
            sanitize_out_option("safe%2Fevil.pdf").as_deref(),
            Some("safe_evil.pdf")
        );
    }

    #[test]
    fn rfc2047_out_decodes_before_sanitize() {
        assert_eq!(
            sanitize_out_option("=?UTF-8?B?0JjQotCe0JPQmCDQm9CU0KMgMjAyNi54bHN4?=").as_deref(),
            Some("ИТОГИ ЛДУ 2026.xlsx")
        );
    }

    // ── Windows reserved names ──────────────────────────────────────
    // The crate replaces reserved names with the replacement string "_".
    // Our wrapper then trims and rejects empty — but "_" is non-empty,
    // so reserved names become "_".  This is safe: "_" is a valid
    // filename on all platforms.

    #[test]
    fn windows_reserved_con_becomes_underscore() {
        assert_eq!(sanitize_out_option("CON").as_deref(), Some("_"));
    }

    #[test]
    fn windows_reserved_nul_txt_becomes_underscore() {
        assert_eq!(sanitize_out_option("NUL.txt").as_deref(), Some("_"));
    }

    #[test]
    fn windows_reserved_com1_becomes_underscore() {
        assert_eq!(sanitize_out_option("com1").as_deref(), Some("_"));
    }

    #[test]
    fn windows_reserved_lpt3_becomes_underscore() {
        assert_eq!(sanitize_out_option("LPT3").as_deref(), Some("_"));
    }

    // ── Trailing dots and spaces ────────────────────────────────────

    #[test]
    fn trailing_dots_stripped() {
        // The crate replaces trailing dots/spaces with replacement "_";
        // our wrapper calls .trim() which handles trailing whitespace.
        // "file.jpg..." → crate → "file.jpg_" → trim → "file.jpg_"
        let result = sanitize_out_option("file.jpg...");
        assert!(result.is_some());
        assert!(result.as_deref().unwrap_or("").starts_with("file.jpg"));
    }

    #[test]
    fn trailing_spaces_stripped() {
        // "file.jpg   " → crate → "file.jpg_" → trim → "file.jpg_"
        // Or our .trim() may catch it. Either way, starts with "file.jpg".
        let result = sanitize_out_option("file.jpg   ");
        assert!(result.is_some());
        assert!(result.as_deref().unwrap_or("").starts_with("file.jpg"));
    }

    // ── Control characters ──────────────────────────────────────────

    #[test]
    fn control_chars_removed() {
        // The crate removes control characters (0x00-0x1F, 0x80-0x9F)
        let result = sanitize_out_option("\x01\x02file.jpg");
        assert!(result.is_some());
        assert!(result.as_deref().unwrap_or("").contains("file.jpg"));
    }

    // ── Normal filenames unmodified ─────────────────────────────────

    #[test]
    fn normal_filename_with_spaces() {
        assert_eq!(
            sanitize_out_option("My Document.pdf").as_deref(),
            Some("My Document.pdf")
        );
    }

    #[test]
    fn extensionless_filename_preserved() {
        assert_eq!(sanitize_out_option("README").as_deref(), Some("README"));
    }

    #[test]
    fn dotfile_preserved() {
        assert_eq!(
            sanitize_out_option(".gitignore").as_deref(),
            Some(".gitignore")
        );
    }
}
